### DISCUSSION TITLE
Add compression as an argument to PdfWriter.write

### DISCUSSION THREAD

**Discussion by pztrick** [py-pdf/pypdf#1087] [Ideas]
Hello,
There is a `PageObject.compressContentStreams` method that no longer seems to be referenced anywhere in the project.

I believe this used to compress PDFs using Zip/Flate to result in a smaller file size.

https://github.com/mstamy2/PyPDF2/blob/8cbca8c8824bcf2ea2b18f9a56e2d7eee3b3fe96/PyPDF2/pdf.py#L2324

---

I tried invoking `page.compressContentStreams` directly before passing to `PdfFileWriter` for outputing to a file stream but it did not seem to work.

Could compression be added to the library? (Ideally as a keyword argument to `PdfFileWriter.write`?)

**Comment by H0L0GH05t** [comment]
+1
Compression is a definite must! I am using PyPDF2 to merge a page I created with reportlab with links onto an existing pdf to add link boxes over certain areas. After merging the link page with the main pdf page, the file size doubles! This is a huge problem for me in terms of performance and download speeds...

**Comment by pztrick** [comment]
@akaiame: This may be useful to you...

In my project I was able to install `imagemagick` which providers the `convert` command-line command. This was the magical incantation I ended up using to reduce my file-size:

```
cmd = "convert -density 130 -compress Zip pdf:- pdf:-"
p = Popen(cmd.split(' '), stdout=PIPE, stdin=PIPE, stderr=PIPE)
ret = p.communicate(input=input_data)[0]
```

I _believe_ it is converting the generated PDF into a 130-dpi image, compressing as Zip, and exporting back to PDF. Note that any rich PDF interactivity like highlighting text would be nixxed as it's now just a 130-dpi image.

cheers..

**Comment by H0L0GH05t** [comment]
@pztrick: I'm already using imagemagick in my project, so I might have to give it a go. I was also considering using ghostscript to downsample after merging the pages, which seems to return the PDF back to its original size before the merge... Still it would be nice to have compression inside PyPDF2!

Thanks for the suggestion!

**Comment by xilopaint** [comment]
Is there any news about this thread?

**Comment by MartinThoma** [comment]
I'm open to add a compression parameter to `PdfWriter.write`. Something like this:

```python
def write(self, stream: StreamType, compress: Literal[None, "FLATE"]=None) -> None:
```

Are there other compression methods supported by PDF?

How do other libraries (PyMuPDF / pikepdf / pdfrw / ...) do it?

As a side-note: https://pypdf2.readthedocs.io/en/latest/user/file-size.html
